[RFC] Clock framework

clementleger opened the pull request:
This serie adds support for a generic clock framework in OP-TEE. This review is mainly a RFC to start discussion on this framework.

Support is split in multiple patches which are the following:
 - Modify driver struct to add a type field needed for clock probing with dt driver
 - Add clock base for framework
 - Add SCMI informations in clock framework (clock id/channel).
 - Add devicetree support to clock framework
 - Add A fixed clock driver which uses this infrastructure
 - Finally, use this clock framework for a generic SCMI clock support

Checkpatch has been applied no the patche but a few warning could remain since code was sometimes not really readable after modifications.

This support is used for upcoming clock support in plat-sam.

etienne-lms commented on core/drivers/scmi-msg/clock_generic.c:
```diff
@@ -0,0 +1,97 @@
+	if (!clk)
+		return SCMI_NOT_FOUND;
+
+	return clk_is_enabled(clk);
```
The scmi should not report the effective clock state, but only the SCMI clock state as seen by the client.
The issue is that client can make unbalanced clock enable/diable requests (i.e. enabling twice a same SMCI clock) so SCMI server must save a boolean state for the clock exposed to ensure client request are always balanced inside OP-TEE core.

clementleger replied:
Ok, I missed that. I will modify that

etienne-lms commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,287 @@
+	void (*disable)(struct clk *clk);
+	int (*is_enabled)(struct clk *clk);
+	int (*set_parent)(struct clk *clk, uint8_t index);
+	uint8_t (*get_parent)(struct clk *clk);
```
Do you have real needs for these set_parent/get_parent handlers?

clementleger replied:
In the current clock tree that I'm porting (at91), there is need to set and get the parent for USB for instance.
We could potentially set them statically but then we will loose the ability to change to new clock rates by using new parents.

etienne-lms commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,287 @@
+ * @scmi_id: SCMI identifier for the clock (used by the generic SCMI clock
+ *	     support)
+ * @scmi_channel_id: SCMI channel for the clock (used by the generic SCMI clock
+ *	     support)
```
these should be outside the clock frameork I think.
SCMI clocks should rely on this clock framework, as a clock consumer like any other clock consumer.

clementleger replied:
Actually, the clock provider set the SCMI ID by itself. I'm not sure who will set the clocks ID if not the provider.
How will the matching be done between hardware clocks and SCMI clocks ID then ?
Currently, the use case is that the clock provider allocate the clock and then sets the SCMI ID to match the hardware clock. But I agree this is not ideal.

clementleger replied:
After thinking a bit more about that, I could add an interface to translate the equivalent of devicetree clock-cells to a struct *clk. The generic scmi clk layer will then export a function to match the clk struct to a scmi_id.
The flow for the platform code would be the following:

``` C
/* Provided by generic clock support */
struct clock_args args = {PMC_TYPE_PERIPHERAL, 42};
clk = clk_get_from_provider("provider_name", args);

/* Provided by generic scmi clk support */
clk_scmi_match(clk, channel_id, scmi_id);
```

Does it sound good to you ?

etienne-lms replied:
I think scmi references should move out of the generic clock framework. If an SCMI server exposes clocks, it should access the clock as any other clock consumer in optee, without specific hooks in the clock framework.

Binding platform clock to SCMI clock is platform specific. What about using a specific node in the DT for defining the clocks exposed, and a dedicated clock consumer driver to exposes these clock to the SCMI server? For example, assuming channel 0 exposes 2 SCMI clocks:
```
scmi-clock-ch0 {
	compatible = "scmi-clock-consumer";
	scmi-channel-id = <0>;

	scmi-clock-ch0@0 {
		reg = <0>;      // the scmi clock ID used to expose the clock
		clocks = <&some_clock>;   // reference to a platform clock
	}
	scmi-clock-ch0@1 {
		reg = <1>;      // the scmi clock ID used to expose the clock
		clocks = <&some_clock_phandle 12>; // reference to another platform clock
	}
}
```

clementleger replied:
That would be a really clean solution ! I like this proposal which is way more flexible than what I have. I'll probably go this way.
Regarding the SCMI part, I did split it completely from the generic clock framework in my last revision.

alexandrebelloni replied:
I'm not sure having a node per clock is a good ID, we explicitely switched away from that in the kernel. Maybe the mapping could be done using a single array of phandles as we know the SCMI IDs start at 0 and are contiguous.

clementleger replied:
Having a single `clocks = <&clock1 arg1 arg 2 &clock2 arg3 arg4 ... >` is probably going to be difficult to read I guess and even harder to modify.
However, I agree that current proposal will yield a consequent number of nodes.

clementleger replied:
On a specific SoC I'm working on, there is 80 clocks that needs to be exposed. Since the properties have the same name, it won't take a lot of space IMHO.
Your solution is more flexible regarding the per clock properties. For instance, allowed SCMI frequency could be specified in theses nodes.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,615 @@
+
+	return TEE_SUCCESS;
+}
+service_init(clk_probe);
```
Could this be probe as early as possible, like early_init

clementleger replied:
Yes it can be.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,615 @@
+	if (!clk)
+		panic();
+
+	memset(clk, 0, alloc_size);
```
prefer `calloc(1, alloc_size)` and remove this `memset()`.

clementleger replied:
Done

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,615 @@
+	}
+}
+
+int clk_register(struct clk *clk)
```
Use `TEE_Result`  type for error code inside OP-TEE core.

clementleger replied:
OK

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,615 @@
+int clk_enable(struct clk *clk)
+{
+	struct clk *parent;
+
```
i think this function and `clk_disable()` miss a reference counter, in case several consumers rely on a single clock.

clementleger replied:
Indeed.

etienne-lms commented on core/drivers/sub.mk:
```diff
@@ -39,5 +39,6 @@ srcs-$(CFG_IMX_RNGB) += imx_rngb.c
 
 subdirs-y += crypto
 subdirs-$(CFG_BNXT_FW) += bnxt
+subdirs-$(CFG_CLK) += clk
```
Maybe rename `CFG_CLK` to an explicit `CFG_CLK_DRIVERS` or `CFG_GENERIC_CLK`?
(platforms can implement clocks without `CFG_CLK` (as they currenly do) so I feel the config switch name a bit confusing).

clementleger replied:
Yes, I'll rename that to CFG_GENERIC_CLK.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,615 @@
+			return NULL;
+
+		clock_cells = fdt_clock_cells(fdt, parent_node);
+		if (clk_idx == 0)
```
i don't really understand the sequence here.

clementleger replied:
Since you can have clocks consuemr using multiple provider, they can have a different number of cells. For instance, you can have the following:

clocks = < &fixed_clk &pmc PMC_TYPE_PERIPHERAL 42 &pmc PMC_TYPE_PERIPHERAL 32>

In order to correctly gather the 3 clock for instance, you have to parse the clock property by progressing of the number of cells defined in the clock provider with #clock-cells. This loop does that, first compute that fixed_clk has 0 cells, add only the phandle size, then compute that the pmc has 2 #clock-cells, etc.

etienne-lms reviewed (commented):
Thanks for this proposal. Some feedback.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,615 @@
+{
+	uint32_t phandle;
+	const uint32_t *prop;
+	int len, idx = 0, parent_node, clock_cells, idx32;
```
Ok

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,287 @@
+ */
+int clk_set_parent(struct clk *clk, unsigned int pidx);
+
+#if defined(CFG_SCMI_MSG_DRIVERS)
```
Agreed, I have moved both SCMI and devicetree related function in separate files.

clementleger commented:
New version which fixes @etienne-lms comments
- Use TEE_Result
- Add enable counter for clocks
- Split devicetree and SCMI code from main clk.c file
- Use 1 variable per line with init
- Add SCMI per clock state
- Use calloc where possible
- Rename CFG_CLK to CFG_GENERIC_CLK
- Use early_init for clock parsing

Additionnaly, add missing bit of code to parse and set assigned-clock* property from device tree.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+	if (clk->flags & CLK_SET_RATE_GATE) {
+		was_enabled = clk_is_enabled(clk);
+		if (was_enabled)
+			clk_disable_no_count(clk);
```
Between here and line 232, you would need to make sure call to `clk_enable()` for that clock will fails with a friendly message.

clementleger replied:
Ok, understood, I fixed it to handle clock re-enabling on failure.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+
+TEE_Result clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	int was_enabled = 0;
```
prefer `bool`

clementleger replied:
Replaced with bool

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+
+	clk->parent = clk->parents[pidx];
+
+	if ((clk->flags & CLK_SET_PARENT_GATE) && was_enabled)
```
testing `was_enabled` is sufficient.

clementleger replied:
Fixed both cases in set_parent and set_rate

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+	/* The parent changed and the rate might also change */
+	clk_compute_rate(clk);
+
+err_out:
```
this is not an error exclusive path, so it should rather be labeled `out:`

clementleger replied:
Renamed err_out label.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+	return ret;
+}
+
+struct clk *clk_get_parent(struct clk *clk)
```
this small function could be defined as `static inline` functions in _clk.h_.

clementleger replied:
Move as a static inline function in clk.h

etienne-lms reviewed (commented):
Comments for "drivers: clk: add generic clock framework"

etienne-lms commented on core/drivers/clk/clk_of.c:
```diff
@@ -0,0 +1,368 @@
+	const struct clk_driver *clk_drv = NULL;
+
+	for_each_dt_driver(drv) {
+		if (drv->type != DT_DRIVER_CLK)
```
here also i think `compatible` is sufficient to find the right driver.

clementleger replied:
Actually, the for_each_dt_driver iterate on all drivers registered in the dt_driver section (which almso contain UART drivers). So we might end up casting the uart driver field into a clk driver field and call probe function.

etienne-lms commented on core/drivers/clk/clk_of.c:
```diff
@@ -0,0 +1,368 @@
+			continue;
+
+		clk_drv = (const struct clk_driver *)drv->driver;
+		for (dm = drv->match_table; dm; dm++) {
```
minor: 
```
for (dm = drv->match_table; dm && dm->compatible; dm++)
	if (!strcmp(dm->compatible, compat))
		return clk_drv;

clementleger replied:
Replace with proposed replacement

etienne-lms commented on core/drivers/clk/clk_of.c:
```diff
@@ -0,0 +1,368 @@
+	};
+}
+
+static int clk_of_set_parent(struct clk *clk, struct clk *parent)
```
prefer `bool` since boolean returned information.

clementleger replied:
Used bool.

etienne-lms reviewed (commented):
Some comments for commit "drivers: clk: add devicetree support". Maybe more to come... not not.
I would also replace the the `OF` reference with a `DT` reference since it's what is it.

clementleger commented on core/kernel/console.c:
```diff
@@ -129,7 +129,7 @@ void configure_console_from_dt(void)
 
 	dt_drv = dt_find_compatible_driver(fdt, offs);
-	if (!dt_drv)
+	if (!dt_drv || dt_drv->type != DT_DRIVER_UART)
```
Agreed, this *should* not return a driver other than uart but if the phandle passed as stdout-path is not pointing to a UART driver (but a clock one for instance) this will crash badly since  there is a cast after that:
```
	sdrv = (const struct serial_driver *)dt_drv->driver;
```
But the driver field depends on the driver type. We could also add more ELF section for each type of driver such as the kernel does.

clementleger replied:
Indeed, for clocks, the check is needed since I want to iterate on the clocks only. As said, another solution would be to have a per driver type sections such as Linux does. It would probably be cleaner but would lead to multiple sections.
Regarding the check for DT_DRIVER_UART, if the devicetree is correct, then yes, there is no need for it, but if the DT is wrong and reference a wrong node instead of a serial then it could be problematic. I'll do as you prefer !

clementleger replied:
Ok, I wanted to make sure this solution was accepted by maintainers.

clementleger commented on core/drivers/clk/fixed_clk.c:
```diff
@@ -0,0 +1,67 @@
+	fcd->rate = fdt32_to_cpu(*freq);
+	clk->priv = fcd;
+
+	if (clk_register(clk)) {
```
I will probably modify the clock setup function to add a return value and panic in the clock framework if setup fails.

clementleger replied:
Added a TEE_Result as a return value from setup. This way, error is forwarded to the upper layer which will panic if needed.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+	free(clk);
+}
+
+static int clk_check(struct clk *clk)
```
Done

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+{
+	struct clk *clk = NULL;
+	unsigned int parent = 0;
+	unsigned int alloc_size = sizeof(struct clk) +
```
Done

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+	if (clk->num_parents > 1) {
+		pidx = clk->ops->get_parent(clk);
+		if (pidx >= clk->num_parents) {
+			EMSG("get_parent returned an invalid value\n");
```
Fixed, remove all \n from messages in clk* files

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+	struct clk *parent;
+
+	if (clk->enabled_count == 0) {
+		EMSG("Unbalanced clk_enable/disable");
```
Assert added.

clementleger commented on core/include/drivers/clk_of.h:
```diff
@@ -0,0 +1,57 @@
+ */
+
+#ifndef CLK_OF_H
+#define CLK_OF_H
```
replace all occurrences of `of` to `dt`

clementleger commented on core/include/drivers/clk_of.h:
```diff
@@ -0,0 +1,57 @@
+struct clk *clk_of_get_by_name(void *fdt, int nodeoffset,
+			       const char *name);
+
+TEE_Result clk_of_register_clk_provider(void *fdt, int nodeoffset,
```
Used clk_dt as prefix

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+	if (clk->enabled_count == ULONG_MAX)
+		return TEE_ERROR_OVERFLOW;
+
+	if (clk->enabled_count) {
```
Added a spinlock and used refcount

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,280 @@
+
+int clk_is_enabled(struct clk *clk)
+{
+	if (clk->ops->is_enabled)
```
Acutally, is_enabled was not used anymore after removed this call. This could be useful later for disabling unused clocks but let this appart for the moment. I removed the is_enabled clock callback.

clementleger commented:
V3:
- Used spinlock & refcount for clock enable/disable
- Rework usage of flags SET_*_GATE (wrong interpretation of their usage), if clock is enabled whil doing these operations, abort operation.
- Use correct types where applicatble (bool, TEE_Result)
- Renamed of -> dt
- Remove SCMI code entirely of this patch, this will be send separately on top of this support
- Remove flawed usage of flag CLK_SET_RATE_PARENT (parent rate was not computed properly)
- Removed is_enabled callback which was unused

clementleger commented:
- Fixed build due to clk_dt_probe being undefined

vesajaaskelainen commented:
We have been looking how the Xilinx Zynq7000 with Linux & OP-TEE could work at least two things might fall in SCMI umbrella:
- Peripheral reset lines
- Clock controls

Both of these require processor to be in secure state.

What are your ideas should this proposed work in this PR be used for this?

At least peripheral reset lines seem to be missing from this set.

If you want to look at Xilinx Zynq7000 TRM:
https://www.xilinx.com/support/documentation/user_guides/ug585-Zynq-7000-TRM.pdf
- 25 Clocks
  - 25.9 Register Overview
- 26 Reset System
  - 26.5.3 Peripheral Reset Control

Example Linux clock driver that would need re-routing:
- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/clk/zynq/clkc.c?h=v5.13

And device tree reference:
- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/boot/dts/zynq-7000.dtsi?h=v5.13#n275

clementleger commented:
> We have been looking how the Xilinx Zynq7000 with Linux & OP-TEE could work at least two things might fall in SCMI umbrella:
> 
> * Peripheral reset lines
> * Clock controls
> 
> Both of these require processor to be in secure state.
> 
> What are your ideas should this proposed work in this PR be used for this?
> 
> At least peripheral reset lines seem to be missing from this set.

This series does not target reset but only clocks. Reset control is out of scope.

> 
> If you want to look at Xilinx Zynq7000 TRM:
> https://www.xilinx.com/support/documentation/user_guides/ug585-Zynq-7000-TRM.pdf
> 
> * 25 Clocks
>   
>   * 25.9 Register Overview
> * 26 Reset System
>   
>   * 26.5.3 Peripheral Reset Control
> 
> Example Linux clock driver that would need re-routing:
> 
> * https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/clk/zynq/clkc.c?h=v5.13
> 
> And device tree reference:
> 
> * https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/boot/dts/zynq-7000.dtsi?h=v5.13#n275

You'll have to port the Linux driver (Be careful about the license) into OP-TEE. The series add a minimal clock framework to handle basic clock trees. More advanced support (CLK_SET_RATE_PARENT for instance) as well as support for generic clock support in SCMI will be added later.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+	if (parent)
+		clk_enable(parent);
+
+	if (clk->ops->enable) {
```
minor:prefer:
```c
	if (clk->ops->enable && clk->ops->enable(clk))
		return TEE_ERROR_GENERIC;

clementleger replied:
Ok, changed it.

etienne-lms commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,202 @@
+ * Returns 0 on success or a negative value on error.
+ */
+
+static inline int clk_is_enabled(struct clk *clk)
```
s/int/bool/

clementleger replied:
Ok, also fixed the returned value.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+	TEE_Result ret = TEE_SUCCESS;
+
+	cpu_spin_lock(&clk->lock);
+	if (rate == clk->rate)
```
lgtm: suggestion (easier to read):
```c
	cpu_spin_lock(&clk->lock);
	if (rate != clk->rate) {
		if (clk->flags & CLK_SET_RATE_GATE && clk_is_enabled(clk))
			ret = TEE_ERROR_BAD_STATE;
		else
			ret = clk_set_rate_no_lock(clk, rate);
	}
	cpu_spin_unlock(&clk->lock);

clementleger replied:
As you wish.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+
+	was_enabled = clk_is_enabled(clk);
+	if (was_enabled)
+		clk_disable_no_lock(clk);
```
this does not ensure the clock is disabled, there maybe other refcounts accumulated. This should return an error, as done in `clk_set_rate()`

clementleger replied:
This call to disable is meant to maintain the refcount to a correct value not to disable the clock. If the clock needs to be disabled while being modified, the driver should handle that. But CLK_SET_RATE_GATE is there for that kind of scenarios where the clock should not be modified with enable peripheral.

etienne-lms replied:
ok, i understand. thanks.
I suggest an short inline comment like `/* Decrement refcounts while on old parent clock */`
and `/* Re-increment refcounts with the new parent clock */` at line 253.

Or maybe no inline comment but rename variable `was_enabled` to `update_refcounts`.

clementleger replied:
I guess the comment is more appropriate here, I'll update that.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+	if (!clk_check(clk))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	SLIST_INSERT_HEAD(&clk_list, clk, link);
```
this list is no more used. can be removed.

clementleger replied:
Indeed.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+	return TEE_SUCCESS;
+}
+
+unsigned long clk_get_rate(struct clk *clk)
```
move to _clk.h_ as `static inline`.

clementleger replied:
Moved.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+
+TEE_Result clk_enable(struct clk *clk)
+{
+	TEE_Result ret;
```
alwyas initialize local variables where defined.
`TEE_Result ret = TEE_ERROR_GENERIC;` is usually a good candidate.

For consistency, optee_os usually use `res` for `TEE_Result` typed variables. I would prefer you use `res` instead of `ret`. Yet, that's not a strong opiion, feel free to discard :)

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+
+static void clk_disable_no_lock(struct clk *clk)
+{
+	struct clk *parent;
```
`= NULL`

clementleger replied:
Missed that one.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+			return TEE_ERROR_GENERIC;
+	}
+
+	clk_compute_rate(clk);
```
could move inside instruction block above.

clementleger replied:
Yes.

etienne-lms reviewed (commented):
Comments for commits :
"drivers: add type field to dt_driver struct"
"drivers: clk: add generic clock framework"

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+
+#include <drivers/scmi.h>
+
+#define SCMI_MAX_CLK_NAME_LEN	16
```
Done

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,202 @@
+#include <sys/queue.h>
+#include <tee_api_types.h>
+
+#define CLK_NO_SCMI_ID	-1
```
Done

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,202 @@
+ * @parent: Current parent
+ * @num_parents: Number of parents
+ * @rate: Current clock rate (cached after init or rate change)
+ * @scmi_id: SCMI identifier for the clock (used by the generic SCMI clock
```
Forgot that, removed it.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+		pidx = clk->ops->get_parent(clk);
+		if (pidx >= clk->num_parents) {
+			EMSG("get_parent returned an invalid value");
+			panic();
```
Fixed

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,368 @@
+	}
+}
+
+TEE_Result clk_dt_probe(void)
```
I think you can discard commit "kernel: boot: probe clocks before console" and call `clk_dt_probe()` from `early_init()` initcall level. There is no need to call it from `boot_init_primary_late()` specifically, do we? Note the console config is done from `boot_init_primary_late()` because it initially only accessed the 'external dt' and every external_dt accessed where called from `boot_init_primary_late()`.

If console driver needs clocks then it's an issue. What about a console driver that would need clock and pin control. and maybe  a reset controller driver. I think it would not be nice to have all these specific drivers initialized from `boot_init_primary_late()` right before `configure_console_from_dt()` is called.
What do you think?

clementleger replied:
Maybe the console are no exception and should be probed later also ? For instance, having an early console is needed to allow debugging but once done, OP-TEE could probably use a more full featured/secure one using devicetree.
But you are right, console probably need (at least) to set pinctrl and clocks to correctly work.

So in the meantime until this is sorted out, I'll reswitch to early_init ;)

jforissier replied:
> Maybe the consoles are no exception and should be probed later also ? For instance, having an early console is needed to allow debugging 

True, that is hard-coded per-platform with various degrees of configure-ability.

> but once done, 
> OP-TEE could probably use a more full featured/secure one using devicetree.

We have it more or less, see [here](https://github.com/OP-TEE/optee_os/commit/4dc31c52544a7241e97d0452d512330ee30ed664) and [here](https://github.com/OP-TEE/optee_os/commit/08baa8c9ae08f355dfd9934a86e8d9dbc913c5a1)). However...

> But you are right, console probably need (at least) to set pinctrl and clocks to correctly work.

...not sure how we would achieve this.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,279 @@
+	struct clk *clk = NULL;
+	unsigned int parent = 0;
+	size_t alloc_size = sizeof(struct clk) +
+				  parent_count * sizeof(struct clk *);
```
Fixed.

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,196 @@
+ * @get_rates_array: Get the supported clock rates as array
+ */
+struct clk_ops {
+	int (*enable)(struct clk *clk);
```
Ok, I changed all ints return values to booleans.

clementleger replied:
I'll use a TEE_Result since it will indeed carry more information and avoid inverting all the return values from my existing clock tree drivers.

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,202 @@
+	const struct clk_ops *ops;
+	struct clk **parents;
+	struct clk *parent;
+	uint8_t num_parents;
```
Ok, I replaced all instances of uint8_t for parent by size_t

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,368 @@
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
```
Removed them.

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,368 @@
+	const struct dt_device_match *dm = NULL;
+	const struct clk_driver *clk_drv = NULL;
+
+	IMSG("Probing clocks from device tree");
```
Used DMSG since this seems like a debug than flow trace.

clementleger commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,111 @@
+	}
+
+/**
+ * clk_dt_get_by_idx - Get a clock at a specific index in "clocks" proeprty
```
Fixed typo.

clementleger commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,111 @@
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the subnode containing a clock property
+ * @clk_idx: Clock index to get
+ * Returns a clk struct matching the clock at index clk_idx in clocks property
```
Ok, also fixed it for clk_dt_get_by_name

etienne-lms replied:
s/clk struct/clk struct reference/

jenswi-linaro commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+ * Copyright (c) 2021, Bootlin
+ */
+
+#ifndef CLK_H
```
The guard should be `__DRIVERS__CLK_H`

clementleger replied:
Ok, I must say the include guards don't seems really homogeneous, I can see the following:
- #define __DRIVERS_GIC_H
- #define SCMI_MSG_H
- #define DRIVERS_WDT_H
Your proposal uses 2 _ after DRIVERS.
Which one should I use ?

jenswi-linaro replied:
Sorry, my bad, I meant `__DRIVERS_CLK_H`. The important part here is a guard that is guaranteed unique without searching through the source tree first. Personally I prefer a `__` prefix in addition to that since it gets out of the way in other aspects too, so that's what I'd suggest.

clementleger replied:
Ok, totally agree.

jenswi-linaro commented on core/drivers/sub.mk:
```diff
@@ -39,5 +39,6 @@ srcs-$(CFG_IMX_RNGB) += imx_rngb.c
 
 subdirs-y += crypto
 subdirs-$(CFG_BNXT_FW) += bnxt
+subdirs-$(CFG_GENERIC_CLK) += clk
```
To me "generic" goes without saying. How about `CFG_DRIVERS_CLK` or `CFG_CORE_CLK` instead?

clementleger replied:
CFG_DRIVERS_CLK_SUPPORT might be better seems it implies that there is some support for clocks ?

jenswi-linaro replied:
OK, sounds good.

jenswi-linaro commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+	const char *name;
+	void *priv;
+	const struct clk_ops *ops;
+	struct clk **parents;
```
We could save a pointer by putting this at the end of this struct instead as:
``` C
struct clk *parents[];
```

clementleger replied:
Yes indeed ! Didn't thought about this one !

jenswi-linaro reviewed (commented):
Comments for the commit "drivers: clk: add generic clock framework"

jenswi-linaro commented:
Do you have any driver using this framework?

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
```
Fixed.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+		      struct clk **parent_clks, size_t parent_count)
+{
+	struct clk *clk = NULL;
+	unsigned int parent = 0;
```
Fixed.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+{
+	unsigned long parent_rate = 0;
+
+	clk->rate = 0;
```
Removed

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+static TEE_Result clk_enable_no_lock(struct clk *clk)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct clk *parent;
```
Fixed

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+		if (clk_is_enabled(clk)) {
+			res = TEE_ERROR_BAD_STATE;
+			goto out;
+		}
```
Ok, I must say I don't find it more readable than previous style but if its ok for you. I've done it.

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+/**
+ * clk_free - Free a clock structure
+ *
+ * @clk: Clock to be freed
```
Ok.

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+ *
+ * @clk: Clock to be set with the rate
+ * @rate: Rate to set
+ * Returns 0 on success or a negative value on error.
```
Indeed. I'll check for other mistakes.

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+ * @clk: Clock to be enabled
+ * Returns 0 on success or a negative value on error.
+ *
+ * Note: This function will also enable parents if needed.
```
Ok

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+ * clk_disable - Disable a clock
+ *
+ * @clk: Clock to be disabled
+ * Returns 0 on success or a negative value on error.
```
Removed returns comment

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+ * @clk: Clock for which the number of parents is needed
+ * Return the number of parents
+ */
+size_t clk_get_num_parents(struct clk *clk);
```
Indeed. Moved it to clk.h

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+			       unsigned long parent_rate);
+	unsigned long (*get_rate)(struct clk *clk,
+				  unsigned long parent_rate);
+	TEE_Result (*get_rates_array)(struct clk *clk, size_t start_index,
```
Yes, I'll add it for SCMI support

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,369 @@
+#include <libfdt.h>
+#include <stddef.h>
+
+struct clk *clk_dt_get_simple_clk(struct clk_dt_phandle_args *args __unused,
```
Done

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,369 @@
+#include <drivers/clk_dt.h>
+#include <kernel/boot.h>
+#include <kernel/panic.h>
+#include <initcall.h>
```
Moved initcall.h above kernel includes

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	cpu_spin_lock(&clk->lock);
```
Out of curiosity, I saw that xsave/xrestore disable/enable exceptions but why is that needed here ?

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,369 @@
+					    const uint32_t *prop)
+{
+	unsigned int arg = 0;
+	struct clk_dt_phandle_args pargs = {0};
```
Ok.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+
+TEE_Result clk_register(struct clk *clk)
+{
+	if (!clk_check(clk))
```
Yes, this can be an assert.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+		break;
+	default:
+		pidx = clk->ops->get_parent(clk);
+		if (pidx >= clk->num_parents)
```
I don't have any idea right now to deal with it cleanly :/ The assert might be the best option I guess.

jenswi-linaro replied:
Agree, this is likely a static configuration error anyway.

clementleger replied:
Done

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+static TEE_Result clk_enable_no_lock(struct clk *clk)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct clk *parent;
```
Done.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+	if (!refcount_dec(&clk->enabled_count))
+		return;
+
+	if (clk->ops->disable)
```
Actually, refcount_dec will assert if the counter value is 0 so I guess this can't be called on an already disabled clock or it will assert. But maybe I missed something.

jenswi-linaro replied:
You're right, nevermind.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+	if (res)
+		return res;
+
+	clk_compute_rate(clk);
```
clk->ops->set_rate is meant to configure the hardware with the requested rate.
clk_compute_rate however, get the rate from the clock and initialize the clk->rate field.

jenswi-linaro replied:
The resulting call to `clk->ops->get_rate()` is a bit disturbing. Surely `clk->ops->set_rate()` can be expected to know (and report back) which rate was set?

jenswi-linaro replied:
OK

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+ * clk_register - Register a clock with the clock framework
+ *
+ * @clk: Clock struct to be registered
+ * Returns 0 on success or a negative value on error.
```
Fixed after @etienne-lms comments.

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+ * @clk: Clock to be enabled
+ * Returns 0 on success or a negative value on error.
+ *
+ * Note: This function will also enable parents if needed.
```
Fixed after @etienne-lms comment.

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,193 @@
+ * clk_disable - Disable a clock
+ *
+ * @clk: Clock to be disabled
+ * Returns 0 on success or a negative value on error.
```
Fixed after @etienne-lms comment.

etienne-lms replied:
remove the line

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+	unsigned long parent_rate = 0;
+
+	if (clk->parent)
+		parent_rate = clk->parent->rate;
```
> Do you have any driver using this framework?

Yes I have a full clock tree ported to this framework. It will be upstreamed in a near future.

clementleger commented:
> Do you have any driver using this framework?

Yes I have a full clock tree ported to this framework. It will be upstreamed in a near future.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+out:
+	/* Call is needed to increment refcount on the new parent tree */
+	if (was_enabled)
+		clk_enable_no_lock(clk);
```
I'm not really sure if we should panic or return an error. But if so, the clock is disabled and the peripheral behind it might not work anymore.

jenswi-linaro commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,263 @@
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	cpu_spin_lock(&clk->lock);
+	if (rate != clk->rate) {
```
If the resulting rate is rounded this test may always be `true`. Perhaps that caller already knows that and always uses a good multiple. I wonder if this test is needed at all. If makes sense to keep it, what about the `if (clk->flags & CLK_SET_RATE_GATE && clk_is_enabled(clk))`, shouldn't that always be tested? I mean it seems to suggest that `clk_set_rate()` was called in a bad state.

clementleger replied:
Indeed, I removed the `rate != clk-rate` check. Maybe I will reintroduce that later when rounding rate operation will be supported to actually check that we really are going to set the same rate.

jenswi-linaro replied:
OK

clementleger commented:
New version:
- Use a global lock for full clock tree
- Renamed CFG_GENERIC_CLK to CFG_DRIVERS_CLK_SUPPORT
- Add new configuration options in config.mk for fixed-clock and dt support
- Add MAINTAINERS entry for clock support

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,299 @@
+}
+
+static struct clk *clk_get_parent_no_lock(struct clk *clk)
+{
```
if this function is not planned to evolve, maybe better use ` = clk->parent` where called.

clementleger replied:
Ok, I removed it since the lock was also removed.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,299 @@
+	parent = clk_get_parent_no_lock(clk);
+	if (parent) {
+		res = clk_enable_no_lock(parent);
+		if (res)
```
decrease refcount on failure

clementleger replied:
Added an error path.

etienne-lms commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,172 @@
+ * clk_set_rate - Set a clock rate
+ *
+ * @clk: Clock to be set with the rate
+ * @rate: Rate to set
```
`Rate to set in Hz`

etienne-lms reviewed (commented):
comments for commit "drivers: clk: add generic clock framework"

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,299 @@
+	 */
+	clk_init_parent(clk);
+
+	cpu_spin_lock(&clk_lock);
```
Agreed that this is not expected to happen and thus, the spinlock is not useful here.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,299 @@
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	cpu_spin_lock(&clk_lock);
```
Ok, thanks for the info, I will switch to using xsave/xrestore variant of spinlock.

jforissier replied:
> You need to mask foreign (non-secure) interrupts 

Yes, and there is an `assert()` in the spinlock code for this very important condition.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,299 @@
+
+	cpu_spin_lock(&clk_lock);
+	rate = clk_get_rate_no_lock(clk);
+	cpu_spin_unlock(&clk_lock);
```
Indeed 32 bits load/store is done in a single operation, no need for that.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,299 @@
+{
+	struct clk *parent = NULL;
+
+	cpu_spin_lock(&clk_lock);
```
Same answer than before, indeed this is not needed.

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,363 @@
+				 const struct clk_driver *clk_drv)
+{
+	int status = 0;
+	TEE_Result res;
```
= TEE_ERROR_GENERIC

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,363 @@
+
+	for (i = 0; i < clk_get_num_parents(clk); i++) {
+		if (clk_get_parent_by_index(clk, i) == parent) {
+			clk_set_parent(clk, i);
```
```c
if (clk_set_parent(clk, i) != TEE_SUCCESS)
	panic(res);

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,363 @@
+
+	rate_prop = fdt_getprop(fdt, nodeoffset, "assigned-clock-rates",
+				&rate_len);
+	if (rate_len)
```
```patch
-	if (rate_len)
		rate_len /= sizeof(uint32_t);

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,363 @@
+
+		parent = clk_dt_get_by_idx_prop("assigned-clock-parents", fdt,
+						nodeoffset, clock_idx);
+
```
remove empty line

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,363 @@
+		if (rate_prop && clock_idx <= rate_len) {
+			rate = fdt32_to_cpu(rate_prop[clock_idx]);
+			if (rate)
+				clk_set_rate(clk, rate);
```
if (clk_set_rate(clk, rate) != TEE_SUCCESS)
	panic();

etienne-lms reviewed (commented):
comments for commit "drivers: clk: add devicetree support".

etienne-lms commented on core/drivers/clk/fixed_clk.c:
```diff
@@ -0,0 +1,66 @@
+
+static TEE_Result fixed_clock_setup(const void *fdt, int offs)
+{
+	const uint32_t *freq;
```
= NULL;
= ...

etienne-lms commented on core/drivers/clk/fixed_clk.c:
```diff
@@ -0,0 +1,66 @@
+#include <kernel/dt.h>
+#include <malloc.h>
+#include <stdint.h>
+#include <util.h>
```
util.h not needed.

etienne-lms commented on core/drivers/clk/fixed_clk.c:
```diff
@@ -0,0 +1,66 @@
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <initcall.h>
+#include <io.h>
```
initcall.h & io.h not needed

clementleger replied:
Also removed assert, kernel/* and assert.

etienne-lms reviewed (commented):
minor comments for commit "drivers: clk: add fixed-clock driver".

clementleger commented:
Fixed two potential problems with assigned-clock parsing:
- clk_dt_set_parent would report a failure due to missing eearly return
- clk_dt_get_by_idx would loop forever due to missing continue

jenswi-linaro commented:
This is over all good, but there is one problem. For memory constrained platforms it uses too much memory. I see a few options:
1. Go ahead with this and try to fix the memory usage later to make it work with memory constrained platforms too
2. Fix the memory usage problem before merging this
3. Go ahead with this and manage clocks by other means on memory constrained platforms

Option 3 is a last resort and a failure in my opinion. OP-TEE should be able to scale down enough to be useful on memory constrained platforms. So the choice is between 1 and 2.

One way of fixing the memory usage is to make `struct clk` smaller, maybe only keeping the `ops` member and then use `container_of()` macro to access a larger struct when needed. Look at the `struct serial_chip` usage for examples.

clementleger commented:
Hi Jens,

> One way of fixing the memory usage is to make struct clk smaller, maybe only keeping the ops member and then use container_of() macro to access a larger struct when needed. Look at the struct serial_chip usage for examples.

Just to clarify, this would only target the `void *priv` field right ? If so, since clk uses flexible member array, this would require the clk struct ot be located at the end of the encompassing structure, but it should be ok.

jenswi-linaro commented:
> Hi Jens,
> 
> > One way of fixing the memory usage is to make struct clk smaller, maybe only keeping the ops member and then use container_of() macro to access a larger struct when needed. Look at the struct serial_chip usage for examples.
> 
> Just to clarify, this would only target the `void *priv` field right ? If so, since clk uses flexible member array, this would require the clk struct ot be located at the end of the encompassing structure, but it should be ok.

I'm afraid I meant a bit more than that. I know this might mean a big change.

> 
> All other fields of the `clk` struct are mandatory for the core. In serial_chip, there is no field used by the core, only the ops, which explains why only the ops are present in the struct `serial_chip`.

Yes, but it's like that by conscious design.

We should perhaps wait for feedback from @etienne-lms and others too, but it's holiday season now so it may take some time.

clementleger commented:
> 
> I'm afraid I meant a bit more than that. I know this might mean a big change.
> 

Could you please be a bit more explicit by describing briefly what you'd expect ?

jenswi-linaro commented:
> > I'm afraid I meant a bit more than that. I know this might mean a big change.
> 
> Could you please be a bit more explicit by describing briefly what you'd expect ?

For good accuracy on requirements I think we need input from @etienne-lms too. So what I say here should be taken with a grain of salt. I'd start with only an `ops` pointer and extending the struct as needed in the callbacks with help of `container_of()`.

What we have here, I assume, covers the generic worst case in complexity. I assume that in many specific cases not so much complexity is needed. Some of the structs may even be kept as global variables in .data or .bss instead of being allocated on the heap. That saves a bit of padding in the heap and also gives a name when doing "info symbol <address>" in gdb. The `name` field is a luxury we usually can do without, etc. Every byte counts, the heap is more expensive than RO-data. There's a tradeoff of course, it still has be be maintainable. Consider a heap size of perhaps 24 kB which is used for more than just clocks and you don't have much to spare.

clementleger commented:
> What we have here, I assume, covers the generic worst case in complexity. I assume that in many specific cases not so much complexity is needed. Some of the structs may even be kept as global variables in .data or .bss instead of being allocated on the heap. That saves a bit of padding in the heap and also gives a name when doing `"info symbol <address>"` in gdb. The `name` field is a luxury we usually can do without, etc. Every byte counts, the heap is more expensive than RO-data. There's a tradeoff of course, it still has be be maintainable. Consider a heap size of perhaps 24 kB which is used for more than just clocks and you don't have much to spare.

Ok, that is more clear, thanks Jens.

A bit out off topic (but not so much actually), regarding memory usage, have you considered discarding init data/text sections after startup ? I did not see such mecanism (marking init functions) but it could actually save quite some space. Note that I'm not aware of all soft/hard limitations, so this question might sound a bit stupid.

jenswi-linaro commented:
> A bit out off topic (but not so much actually), regarding memory usage, have you considered discarding init data/text sections after startup ? I did not see such mecanism (marking init functions) but it could actually save quite some space. Note that I'm not aware of all soft/hard limitations, so this question might sound a bit stupid.

With paging enabled, which is our main low memory user case get this automatically. The unused init pages will at some point be reused. It still uses some memory but it's only 32 bytes per 4 kB page so it's not too bad.

ldebieve commented on core/include/kernel/dt.h:
```diff
@@ -52,8 +52,13 @@ struct dt_device_match {
 };
 
+enum dt_driver_type {
+	DT_DRIVER_UART,
```
Could it be useful to add a DT_DRIVER_NOTYPE = 0 that will allow to define a dt_driver_type without a type defined?
It could be useful to loop around a dt_driver list to probe them and avoid a mistake that the first define type is UART.

clementleger replied:
I will add that yes.

etienne-lms commented:
From https://github.com/OP-TEE/optee_os/pull/4705#issuecomment-886838633

> > A bit out off topic (but not so much actually), regarding memory usage, have you considered discarding init data/text sections after startup ? (...)
>
> With paging enabled, which is our main low memory user case get this automatically. The unused init pages will at some point be reused. It still uses some memory but it's only 32 bytes per 4 kB page so it's not too bad.

I fully agree. Moreover, th clock framework Initialization will start once pager is up and running, so the init is fully pageable, and that's fine.

etienne-lms commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,172 @@
+ * @pidx: parent index to set
+ * Returns a TEE_Result compliant value
+ */
+TEE_Result clk_set_parent(struct clk *clk, size_t pidx);
```
In reply to comment [886819931](https://github.com/OP-TEE/optee_os/pull/4705#issuecomment-886819931):
> For good accuracy on requirements I think we need input from @etienne-lms too. (...)

All the generic things a clk driver layer can provide may be duplication of how the platform clock driver describes the hardware. I was thinking platform backend driver that do it its own way.

Here, it could be a config switch `CFG_DRIVERS_PLAT_CLK=y|n` using _clk.h_ as the shim to a `plat_clk_*()` platform driver interface.
```patch
@@@
 #ifndef __DRIVERS_CLK_H
 #define __DRIVERS_CLK_H
 
+#ifndef CFG_DRIVERS_PLAT_CLK
 #include <kernel/refcount.h>
 #include <stdint.h>
 #include <tee_api_types.h>

@@@
  * Returns a TEE_Result compliant value
  */
 TEE_Result clk_set_parent(struct clk *clk, size_t pidx);
+#else /* !CFG_DRIVERS_PLAT_CLK */
+/*
+ * CFG_DRIVERS_PLAT_CLK relies on a single memory optimized
+ * platform clock provider. Some features from clk.h are not supported,
+ * mainly clock parenthood description and configuration.
+ */
+
+/* Opaque reference struct clk, meaninful to plat_clk_*() functions */
+struct clk;
+
+/* Platform shall implement to following functions */
+char *plat_clk_get_name(struct clk *clk);
+struct clk *plat_clk_find_by_name(const char *name);
+bool plat_clk_is_valid(struct clk *clk);
+unsigned long plat_clk_get_rate(struct clk *clk);
+TEE_Result plat_clk_change_rate(struct clk *clk, unsigned long rate);
+TEE_Result plat_clk_enable(struct clk *clk);
+void plat_clk_disable(struct clk *clk);
+
+/* Clock driver helper functions route to plat_clk_*() when supported */
+static inline const char *clk_get_name(struct clk *clk)
+{
+	return plat_clk_get_name(clk);
+}
+
+static inline struct clk *clk_alloc(const char *name,
+				    const struct clk_ops *ops __maybe_unused,
+				    struct clk **parent_clks __maybe_unused,
+				    size_t parent_count __maybe_unused)
+{
+	assert(!ops && !parent_clks && !parent_count);
+
+	return plat_clk_find_by_name(name);
+}
+
+static inline void clk_free(struct clk *clk __unused)
+{
+}
+
+TEE_Result clk_register(struct clk *clk)
+{
+	if (plat_clk_is_valid(clk))
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+}
+
+static inline unsigned long clk_get_rate(struct clk *clk)
+{
+	return plat_clk_get_rate(clk);
+}
+
+static inline TEE_Result clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	return plat_clk_change_rate(clk, rate);
+}
+
+static inline TEE_Result clk_enable(struct clk *clk);
+{
+	return plat_clk_enable(clk);
+}
+
+static inline struct clk *clk_get_parent_by_index(struct clk *clk __unused,
+						  size_t pidx __unused);
+{
+	// maybe add plat_clk_parent_*() functions in plat_clk.h
+	return NULL;
+}
+
+static inline TEE_Result clk_set_parent(struct clk *clk__unused,
+					size_t pidx __unused);
+{
+	DMSG("Unpexpected call with CFG_DRIVERS_PLAT_CLK");
+	return TEE_ERROR_GENERIC;
+}
+
+static inline void clk_disable(struct clk *clk);
+{
+	plat_clk_disable(clk);
+}
+
+static inline struct clk *clk_get_parent(struct clk *clk __unused)
+{
+	return NULL;
+}
+
+static inline size_t clk_get_num_parents(struct clk *clk __unused)
+{
+	return 0;
+}
+#endif /* !CFG_DRIVERS_PLAT_CLK */
-
 #endif /* __DRIVERS_CLK_H */
```

But I think this approach is why @jenswi-linaro mentioned as the failure option 3 in https://github.com/OP-TEE/optee_os/pull/4705#issuecomment-886414835.

clementleger commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,103 @@
+ * struct clk_driver - clk driver setup struct 
+ * setup:
+ */
+struct clk_driver {
```
I went with what already exist for the serial: `serial_driver`. I'm ok to rename it but it will be less coherent with existing things.

clementleger commented on mk/config.mk:
```diff
@@ -685,4 +685,7 @@ endif
\ No newline at end of file
+# CFG_DRIVERS_CLK_DT_SUPPORT embeds devicetree clock parsing support
+CFG_DRIVERS_CLK_SUPPORT ?= n
+CFG_DRIVERS_CLK_DT_SUPPORT ?= $(call cfg-all-enabled,CFG_DRIVERS_CLK_SUPPORT CFG_DT)
```
@jenswi-linaro was ok with that. I'm OK to rename it, but it could be nice if everybody was ok to agree first before that :)

jenswi-linaro replied:
I'm fine with what @etienne-lms is suggesting.
By the way, I can't find the comment you're referring to. Not that it matters much now though.

clementleger replied:
So lets go with  CFG_DRIVERS_CLK_DT if you are ok
@jenswi-linaro : https://github.com/OP-TEE/optee_os/pull/4705#discussion_r664327724

etienne-lms replied:
I think it useless to add _SUPPORT suffix for a y/n config switch that obviously enable some kind of support.

I would be i favor to:
CFG_DRIVERS_CLK
CFG_DRIVERS_CLK_DT
CFG_DRIVERS_CLK_FIXED

clementleger replied:
Ok, done !

clementleger commented on mk/config.mk:
```diff
@@ -681,3 +681,8 @@ ifeq (,$(CFG_HWRNG_QUALITY))
+# When enabled, CFG_DRIVERS_CLK_SUPPORT embeds a clock framwork in OP-TEE core.
+# This clock framework allows to describe clock tree and provide functions to
+# get and configure the clocks.
+CFG_DRIVERS_CLK_SUPPORT ?= n
```
Ditto

clementleger replied:
Also done

clementleger commented on core/drivers/clk/fixed_clk.c:
```diff
@@ -0,0 +1,60 @@
+
+	clk = clk_alloc(name, &fixed_clk_clk_ops, NULL, 0);
+
+	fcd = malloc(sizeof(struct fixed_clock_data));
```
Ok

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,364 @@
+{
+	struct clk_dt_provider *prv = NULL;
+
+	prv = malloc(sizeof(*prv));
```
Ok

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,364 @@
+	}
+
+	clk_probe_assigned(fdt, -1);
+
```
once clock drivers are all probed, can't we free the memory allocate for clock provider registering, `struct clk_dt_provider`? I think there are not required anymore, are they?

clementleger replied:
Actually, the provider are still used for `clk_dt_get_by_idx_prop` which is used by drivers to query clocks from device tree. However, providers could probably be freed at the end of OP-TEE boot with a `boot_final` initcall such as `release_external_dt`. I don't think we can expect drivers to query clocks after that.

etienne-lms replied:
agree

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,364 @@
+	while (node >= 0) {
+		status = _fdt_get_status(fdt, node);
+		if (!(status & DT_STATUS_OK_SEC))
+			continue;
```
Indeed, fixed.

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,364 @@
+			continue;
+
+		clk_drv = (const struct clk_driver *)drv->driver;
+		for (dm = drv->match_table; dm && dm->compatible; dm++)
```
Yes, compatible search order is totally wrong and does not match what should be expected. I 'll rework the probe order to parse DT nodes first and match nodes compatible with a clock driver.

clementleger commented:
New version:
- Rework probe order for clocks (DT nodes compatible first instead of driver first)
- Add a boot_final initcall to free clock providers which are not used anymore.
- Fix for checkpatch

etienne-lms commented on core/drivers/clk/fixed_clk.c:
```diff
@@ -0,0 +1,61 @@
+
+	fcd = calloc(1, sizeof(struct fixed_clock_data));
+	if (!clk || !fcd)
+		return TEE_ERROR_OUT_OF_MEMORY;
```
must free the allocated resource before returning. Or could panic()!

Ditto for the other error cases below (lines 48, 55 and 57).

clementleger replied:
Added free/clk_free

etienne-lms commented on MAINTAINERS:
```diff
@@ -158,6 +158,11 @@ F:	core/arch/arm/plat-imx/
 F:	core/arch/arm/plat-imx/registers
 F:	core/drivers/crypto/caam/
 
+Clock driver interface
```
i would replace `interface` with `framework`. (not a strong opinion)

etienne-lms reviewed (commented):
in commit log of commit "drivers: clk: add devicetree support":
- fix CFG_DRIVERS_CLK_DT_SUPPORT into CFG_DRIVERS_CLK_DT
- Replace the `xxx_of_xxx` reference with `_dt_`: clock_dt_get_by_{name|idx}()`.

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,398 @@
+						nodeoffset, clock_idx);
+		if (parent) {
+			if (!clk_dt_set_parent(clk, parent))
+				return;
```
Added error message + panic

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,398 @@
+	return clk_dt_get_by_idx_prop("clocks", fdt, nodeoffset, clk_idx);
+}
+
+static TEE_Result clk_probe_clock_provider_node(const void *fdt, int node);
```
Ok

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,172 @@
+ *
+ * @clk: Clock for which the parent is needed
+ * @pidx: parent index for the clock
+ * Return the clock parent at index pidx or NULL if out of bound
```
Fixed

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,172 @@
+	void *priv;
+	const struct clk_ops *ops;
+	struct clk *parent;
+	size_t num_parents;
```
Done

clementleger commented on core/include/drivers/clk.h:
```diff
@@ -0,0 +1,172 @@
+/**
+ * clk_disable - Disable a clock
+ *
+ * @clk: Clock to be disabled
```
Fixed

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,398 @@
+	return TEE_SUCCESS;
+}
+
+boot_final(clk_dt_release_provider);
```
Ok, switched to driver_init_late, hoping nobody uses it to "really" probe drivers init with clock request (at least currently, no driver should have this).

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,398 @@
+
+	idx = fdt_stringlist_search(fdt, nodeoffset, "clock-names", name);
+	if (idx < 0)
+		return -1;
```
Merged indeed.

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,398 @@
+	SLIST_FOREACH(prv, &clk_dt_provider_list, link) {
+		if (prv->nodeoffset == nodeoffset)
+			return prv;
+	}
```
Ok

clementleger commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,104 @@
+struct clk_dt_phandle_args {
+	int nodeoffset;
+	int args_count;
+	uint32_t args[MAX_CLK_PHANDLE_ARGS];
```
Done

clementleger commented:
New version:
- Fixed commit messages after reviews
- Fixed after @etienne-lms review

jforissier commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,395 @@
+		if (!compat)
+			return TEE_ERROR_GENERIC;
+
+		clk_drv = clk_get_compatible_driver(compat);
```
Wouldn't `dt_find_compatible_driver(fdt, node)` give the same result?

clementleger replied:
Let's say that your node as the following compatible in your node: `compatible = "compat1", "compat2"`
If the driver array contains the driver for compat2 before the driver for compat1 then the order of compatible would not be honored. I guess this should probably be fixed in `dt_find_compatible_driver`. @etienne-lms weren't you talking about that in some comment ?

etienne-lms replied:
Yes, `dt_find_compatible_driver()` does not take into account the ordering in property compatible.
Maybe `clk_get_compatible_driver()` could be changed to use a fixed implementation of `dt_find_compatible_driver()`.

jforissier commented on core/include/kernel/dt.h:
```diff
@@ -52,8 +52,15 @@ struct dt_device_match {
+
 struct dt_driver {
 	const char *name;
+	enum dt_driver_type type;
```
It is not clear to me why this new field is needed. I can see how it can optimize the traversal of the driver list, but aren't the `compatible` strings enough to uniquely match a driver?

clementleger replied:
The problem is that driver are found using the following:

```
	dt_drv = dt_find_compatible_driver(fdt, offs);
	if (!dt_drv || dt_drv->type != DT_DRIVER_UART)
		goto out;

	sdrv = (const struct serial_driver *)dt_drv->driver;
	if (!sdrv)
		goto out;
```
As you may see, the `driver` field is cast into a "serial/clock/whatever" specific structure. So, if the node compatible for a clock point to a serial driver for instance, we'll cast it into the wrong type and probably cause havoc.

jforissier replied:
Ah, got it, sorry I read the old code a bit too quickly. Indeed this could be a problem in case the code is fed with a bad DT.
Could you please clarify in the commit description of "drivers: add type field to dt_driver struct"? My proposal: "...which will allow differentiating drivers when iterating on entries **and thus avoid casting struct dt_driver::driver to the wrong type in case a bad DT is provided**".

clementleger replied:
Yep exactly, wrong DT could cause cast error (I agree that this is very unlikely but better be safe than sorry). Ok for the commit description.

jforissier reviewed (commented):
- For commit " drivers: clk: add devicetree support"
  In the commit description, s/CLK_OF_DECLARE/CLK_DT_DECLARE/
- For commit "drivers: clk: add fixed-clock driver":
  In the commit description, s/CFG_DRIVERS_CLK_DT_SUPPORT/CFG_DRIVERS_CLK_DT/
- For "MAINTAINERS: add clock driver interface entry":
  `Acked-by: Jerome Forissier <jerome@forissier.org>`

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,273 @@
+	clk_init_parent(clk);
+	clk_compute_rate_no_lock(clk);
+
+	DMSG("Registered clock %s, freq %ld", clk->name, clk_get_rate(clk));
```
s/%ld/%lu/ since `unsigned long`.

clementleger replied:
Fixed.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,273 @@
+	if (clk->ops->enable) {
+		res = clk->ops->enable(clk);
+		if (res)
+			return res;
```
Shouldn't we disable the parent clock that was enabled above? (I don't remember if we already discussed that).

clementleger replied:
Hum, I don't think we talked about it but yes, I'll add parent disabling in case of error.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,273 @@
+out:
+	/* Call is needed to increment refcount on the new parent tree */
+	if (was_enabled)
+		res = clk_enable_no_lock(clk);
```
I thins should this restore the original parent clock upon failure.

clementleger replied:
After reviewing this part with some hindsight, I'm thinking the pidx parameter for the parent is not really convenient at all. I will probably replace this with the parent clock struct directly (at least for the public part). For the private interface with clocks, it makes more sense since this is often an id that is used.

clementleger replied:
Fixed

etienne-lms commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,104 @@
+ * setup:
+ */
+struct clk_driver {
+	TEE_Result (*setup)(const void *fdt, int nodeoffset);
```
I think this callback could move to `struct dt_driver`, as a probe-from-dt-node callback reference.
Maybe in a later change...
Note that 'probe` maybe be a better name.

clementleger replied:
Agreed, this would make sense to be in the dt_driver struct. The next serie I have regarding SCMI clocks uses the same mechanism with a generic DT probe. Would you mind waiting for the next serie to introduce that ?

etienne-lms replied:
Ok. Let's see that later.
I played with your clk dt stuff and prepared also some dt_driver probing thing. Hence the many catches i reported.  I'll also create a RFC for my proposal, once your changes are accepted. We discuss by then. @ldebieve you may be interested by these proposals.

clementleger replied:
Ok great ! I'll see if it fits what I need but I guess so.

etienne-lms commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,104 @@
+ * CLK_DT_DECLARE - Declare a clock driver
+ * @__name: Clock driver name
+ * @__compat: Compatible string
+ * @__init: Clock setup function
```
`__setup` would be better argument name IMO.

clementleger replied:
Yes, fixed.

etienne-lms commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,104 @@
+	}; \
+	static const struct dt_device_match __name ## _match_table[] = { \
+		{ .compatible = __compat }, \
+		{ 0 } \
```
prefer `{ }` over `{ 0 }`.

clementleger replied:
Ok

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,395 @@
+		clk_probe_assigned(fdt, child);
+
+		status = _fdt_get_status(fdt, child);
+		if (!status)
```
prefer explicit `if (status == DT_STATUS_DISABLED)`
or even:
```c
	fdt_for_each_subnode(child, fdt, parent_node) {
		clk_probe_assigned(fdt, child);

		if (_fdt_get_status(fdt, child) != DT_STATUS_DISABLED &&
		    fdt_getprop(fdt, child, "assigned-clocks", &len))
			parse_assigned_clock(fdt, child);
	}

clementleger replied:
I'll use `if (status == DT_STATUS_DISABLED)`. I don't think that putting everything under the following `if` eases reading.

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,395 @@
+
+	fdt_for_each_subnode(child, fdt, parent_node) {
+		clk_probe_clock_provider_node(fdt, child);
+		clk_probe_node(fdt, child);
```
Prefer skipping disabled nodes here.

clementleger replied:
Acked

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,395 @@
+	int child = 0;
+
+	fdt_for_each_subnode(child, fdt, parent_node) {
+		clk_probe_clock_provider_node(fdt, child);
```
`clk_probe_clock_provider_node()` return code should checked, at least asserted against unexpected values:
```c
		res = clk_probe_clock_provider_node(fdt, child);
		assert(res == TEE_SUCCESS || res == TEE_ERROR_ITEM_NOT_FOUND);

clementleger replied:
Indeed

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,395 @@
+		if (clk_get_parent_by_index(clk, i) == parent) {
+			if (clk_set_parent(clk, i) != TEE_SUCCESS)
+				panic();
+			return true;
```
nitpick: prefer an empty line above. Yet, that's not a strong opinion.

clementleger replied:
Done

etienne-lms commented on core/drivers/clk/fixed_clk.c:
```diff
@@ -0,0 +1,76 @@
+	res = clk_dt_register_clk_provider(fdt, offs, clk_dt_get_simple_clk,
+					   clk);
+	if (!res)
+		return 0;
```
`return TEE_SUCCESS`

minor: I would suggest the following:
```c
	res = clk_dt_register_clk_provider(fdt, offs, clk_dt_get_simple_clk,
					   clk);
out:
	if (res) {
		free(fcd);
		clk_free(clk);
	}

	return res;
```

clementleger replied:
Wasn't sure of that for the end check. In some way, your suggestion duplicates the existing code for freeing but my `(!res)` is less intuitive.

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,273 @@
+	free(clk);
+}
+
+static bool clk_check(struct clk *clk)
```
add `__maybe_unused`

clementleger replied:
Ok, didn't think of that when assert are disabled.

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,395 @@
+	return TEE_ERROR_GENERIC;
+}
+
+static void clk_probe_node(const void *fdt, int parent_node);
```
can remove this declaration

clementleger replied:
Fixed

clementleger commented:
New version:
- Fix @etienne-lms comments.
- Change clk_set_parent() parent argument to be a pointer instead of an idex (simplifies use)

etienne-lms commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,305 @@
+			pres = clk_get_parent_idx(clk, prev_parent, &prev_pidx);
+			assert(pres == TEE_SUCCESS);
+			pres = clk->ops->set_parent(clk, prev_pidx);
+			assert(pres == TEE_SUCCESS);
```
It now seems to me that `assert(clk_enable_no_lock(clk));` at line 273 would given the same confidence level with less code. What do you think.?

etienne-lms replied:
I mean... this enables a clock that was enabled not so long ago, so if a previous call to clk_enable() succeeded in the past, there is no reason why it would fail on that same clock reference now.

Sorry for this comment after I've asked you to restore the parent clock. Maybe you latest changes introducing `clk_get_parent_idx()` were not really needed.

clementleger replied:
> It now seems to me that assert(clk_enable_no_lock(clk)); at line 273 would given the same confidence level with less code. What do you think.?
Calling a function inside assert is probably not a good idea, when compiled with NDEBUG, the call will be left out.
However, I'm not sure of that assumption for the confidence of the call to succeed. Let's say we switched the parent to another one and the clock is a PLL that needs to lock to the correct frequency. If incorrectly setup, the clock might never lock, so I guess there is still potentially problems for that. I'm not sure however if the wait for the PLL locking should be done in `set_parent()` callback or in the `enable()`.

Anyway, I'm ok that we could consider that `clk_enable_no_lock()` should succeed. However, I'll be more in favor of a `panic()` rather than an `assert()` for the same reason that said before (assert discarded when `CFG_TEE_CORE_DEBUG` is not set to y).

> Sorry for this comment after I've asked you to restore the parent clock. Maybe you latest changes introducing clk_get_parent_idx() were not really needed.

Don't worry, this function was actually previously present in clk_dt.c and has been removed thanks to the modification of clk_set_parent arguments. I also really think it make more sense for the public interface to use a `struct clk *parent` rather than a index.

clementleger replied:
I decided to add a `if (res) panic()`after clock enabling. IMHO, this "should" not happen but let's be cautious.

etienne-lms reviewed (commented):
Comments for commit "drivers: clk: add generic clock framework".

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,305 @@
+					 size_t pidx)
+{
+	__maybe_unused TEE_Result pres = TEE_ERROR_GENERIC;
+	TEE_Result res = TEE_ERROR_GENERIC;
```
Why ? I wanted to make sure we return the correct `res` value in error path (line 275) and that's why I added pres for parent result checking.

etienne-lms replied:
Sorry, I messed-up `res` and `pres`. Maybe move `pres` to the instruction block were used.

clementleger replied:
Ok, but as you said in a later commit, if checks are removed and we try to panic if enable fail, this variable will be useless so I might as well removed it entirely.

etienne-lms replied:
sure :)

clementleger replied:
Ok, I finally removed them.

clementleger commented on core/drivers/clk/clk.c:
```diff
@@ -0,0 +1,305 @@
+
+TEE_Result clk_register(struct clk *clk)
+{
+	assert(clk_check(clk));
```
Not sure this is more readable. The suggestion you made does also the wrong checks and assert with correct clocks.
The correct assert should be the following (I guess ?):

```
	assert(clk && clk->ops &&
	       (!clk->ops->set_parent || clk->ops->get_parent) &&
	       (clk->num_parents <= 1 || clk->ops->get_parent));
```

Anyway, the generated code with `NDEBUG` will be same (ie nothing) but I find the `clk_check()` version more readable and maintainable.

etienne-lms replied:
ok

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for the whole series with the 3 remaining coding style issues fixed (80char/line max).

I think the IBART CI failure is unrelated to this P-R. I've tested qemu/armv8 locally, no issue found.
```
  3962:  pkcs11_1008.8 FAILED first error at pkcs11_1000.c:1737
  3964:  pkcs11_1009.1 FAILED first error at pkcs11_1000.c:2035
```

jforissier commented:
> I think the IBART CI failure is unrelated to this P-R. I've tested qemu/armv8 locally, no issue found.

Yes, most probably this branch is lagging too much behind master, @clementleger please rebase while you're at it then hopefully we can have a clean CI status.

etienne-lms reviewed (commented):
Hello Clément,
I had forgotten to post these 2 comments, will help CI.

Could you rebase your changes on master? I think they are mature to consider merging.

Following our discussions during the review, I have a series of proposals around clocks and driver probing. I was waiting for this P-R to be merged before creating some RFC P-Rs. In the mean time, if interested, you can have a look it: 2 alternates to lower memory footprint (https://github.com/etienne-lms/optee_os/pull/27 and https://github.com/etienne-lms/optee_os/pull/28) and a driver probing proposal (https://github.com/etienne-lms/optee_os/pull/27).

clementleger commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,386 @@
+}
+
+TEE_Result clk_dt_register_clk_provider(const void *fdt, int nodeoffset,
+					struct clk *(get_of_clk)(struct clk_dt_phandle_args *, void *),
```
Fixed !

jforissier commented on mk/config.mk:
```diff
@@ -718,6 +718,8 @@ CFG_PREALLOC_RPC_CACHE ?= y
 CFG_DRIVERS_CLK ?= n
 CFG_DRIVERS_CLK_DT ?= $(call cfg-all-enabled,CFG_DRIVERS_CLK CFG_DT)
 $(eval $(call cfg-depends-all,CFG_DRIVERS_CLK_DT,CFG_DRIVERS_CLK CFG_DT))
+CFG_DRIVERS_CLK_FIXED ?= $(CFG_DRIVERS_CLK_DT)
```
Shouldn't `CFG_DRIVERS_CLK_FIXED` depend on `CFG_DRIVERS_CLK_DT`?
```
$(eval $(call cfg-depends-all,CFG_DRIVERS_CLK_FIXED,CFG_DRIVERS_CLK_DT))
```

etienne-lms replied:
you're right.

Side comment: to ease readibility, maybe we could add all the dependency tested after default config settings:
```
CFG_DRIVERS_CLK ?= n
CFG_DRIVERS_CLK_DT ?= $(call cfg-all-enabled,CFG_DRIVERS_CLK CFG_DT)
CFG_DRIVERS_CLK_FIXED ?= $(CFG_DRIVERS_CLK_DT)

$(eval $(call cfg-depends-all,CFG_DRIVERS_CLK_DT,CFG_DRIVERS_CLK CFG_DT))
$(eval $(call cfg-depends-all,CFG_DRIVERS_CLK_FIXED,CFG_DRIVERS_CLK_DT))
``

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,386 @@
+static SLIST_HEAD(, clk_dt_provider) clk_dt_provider_list =
+				SLIST_HEAD_INITIALIZER(clk_dt_provider_list);
+
+static uint32_t fdt_clock_cells(const void *fdt, int nodeoffset)
```
s/uint32_t/int/

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,386 @@
+	if (len != sizeof(*c))
+		return -FDT_ERR_BADNCELLS;
+
+	return fdt32_to_cpu(*c);
```
cast to `int`.

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,386 @@
+	prv->get_of_clk = get_of_clk;
+	prv->data = data;
+	prv->nodeoffset = nodeoffset;
+	prv->clock_cells = fdt_clock_cells(fdt, nodeoffset);
```
must check to return value is positive

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,386 @@
+		if (res)
+			panic("Failed to setup parent clock");
+
+		clock_cells = fdt_clock_cells(fdt, parent_node);
```
cehck return value is positive

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,386 @@
+	SLIST_FOREACH(prv, &clk_dt_provider_list, link) {
+		if (prv->phandle == phandle)
+			return prv;
+	}
```
nitpicking: can remove the braces (as done around lines 82-84).

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,386 @@
+		return TEE_ERROR_GENERIC;
+
+	for (idx = 0; idx < count; idx++) {
+		compat = fdt_stringlist_get(fdt, node, "compatible", idx,
```
minor: fits in a single line

etienne-lms commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,105 @@
+#include <stdint.h>
+#include <sys/queue.h>
+
+#define MAX_CLK_PHANDLE_ARGS 3
```
can be removed.

etienne-lms commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,105 @@
+/**
+ * struct clk_dt_phandle_args - Devicetree clock args
+ * @nodeoffset: Clock consumer node offset
+ * @args_count: Count of cells for the clock
```
maybe replace `Count of cells for the clock` with `Count of items in @args`

etienne-lms commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,105 @@
+};
+
+/**
+ * struct clk_driver - clk driver setup struct
```
s/clk/Clock/

etienne-lms commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,105 @@
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the subnode containing a clock property
+ * @name: Clock name to get
+ * Returns a clk struct matching the name in "clock-names" property or NULL if
```
s/clk struct/clk struct reference/ or "clk struct pointer".

jenswi-linaro commented:
One option could be to typedef the function pointer. I guess we'll just replace the warning, but at least it will be more readable.

clementleger commented:
I decided to use a typedef to solve this. This is also more maintainable since this definition was used elsewhere.

etienne-lms commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,114 @@
+			       const char *name);
+
+/**
+ * clk_dt_get_fn - Typdef of function to get clock from devicetree properties
```
s/Typdef/Type/

etienne-lms commented on core/include/drivers/clk_dt.h:
```diff
@@ -0,0 +1,114 @@
+ * clk_dt_get_fn - Typdef of function to get clock from devicetree properties
+ *
+ * @args: Pointer to devicetree description of the clock to parse
+ * @data: Pointer to data given at clk_dt_register_clk_provider call
```
Prefer with parentheses to highlight its a function:
```
 * @data: Pointer to data given at clk_dt_register_clk_provider() call

etienne-lms commented on core/drivers/clk/clk_dt.c:
```diff
@@ -0,0 +1,392 @@
+					clk_dt_get_fn get_dt_clk, void *data)
+{
+	struct clk_dt_provider *prv = NULL;
+	int clock_cells;
```
= 0;
